File: peace/gui/src/org/peace_tools/workspace/FWAnalyzer.java
//--------------------------------------------------------------------
//
// This file is part of PEACE.
// 
// PEACE is free software: you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// PEACE is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with PEACE.  If not, see <http://www.gnu.org/licenses/>.
// 
// Miami University makes no representations or warranties about the
// suitability of the software, either express or implied, including
// but not limited to the implied warranties of merchantability,
// fitness for a particular purpose, or non-infringement.  Miami
// University shall not be liable for any damages suffered by licensee
// as a result of using, result of using, modifying or distributing
// this software or its derivatives.
//
// By using or copying this Software, Licensee agrees to abide by the
// intellectual property laws, and all other applicable laws of the
// U.S., and the terms of GNU General Public License (version 3).
//
// Authors:   Dhananjai M. Rao          raodm@muohio.edu
//
//---------------------------------------------------------------------

package org.peace_tools.workspace;

import java.io.PrintWriter;
import org.w3c.dom.Element;

/**
 * A simple class to encapsulate information about the frame/word
 * analyzer that was used to generate Minimum Spanning Tree (MST) data.
 * This class provides the necessary helper methods to marhsall and 
 * unmarshall the needed data from-and-to XML.
 */
public class FWAnalyzer {
	/**
	 * The different kinds of frame/word analyzers that are currently
	 * supported in PEACE. This enumeration provides a mechanism to
	 * consistently refer to one of the algorithms used for computing
	 * distance/similarity metric.
	 */
	public enum FWAnalyzerType {
		TWOPASSD2, D2, D2ZIM, CLU
	};

	/**
	 * Helper method to utilize data from a DOM tree to create a suitable
	 * FWAnalyzer entry. This method is typically  used to create a suitable
	 * analyzer entry when loading a Work space into the GUI.
	 * 
	 * @param analyzer The DOM element to be used for creating the Job
	 * entry and populating with the needed data.
	 * 
	 * @return The newly created analyzer entry based on the DOM data.
	 * 
	 * @throws Exception This method throws an exception when errors occur
	 * during reading and processing elements from the DOM node.
	 */
	public static FWAnalyzer create(Element analyzer) throws Exception {
		// First extract the necessary information from the DOM tree.
		String metric   = analyzer.getAttribute("metric");
		String typeStr  = DOMHelper.getStringValue(analyzer, "Type");
		int    frame    = DOMHelper.getIntValue(analyzer, "WindowSize");
		int    word     = DOMHelper.getIntValue(analyzer, "WordSize");
		String cache    = DOMHelper.getStringValue(analyzer, "CacheType");
		int   cacheSize= DOMHelper.getIntValue(analyzer, "CacheSize");
		// Translate the typeStr to an enumeration value.
		typeStr = typeStr.toUpperCase();
		FWAnalyzerType type = FWAnalyzerType.valueOf(FWAnalyzerType.class, typeStr);
		boolean isDistance  = "distance".equals(metric);
		// Now that we have sufficient information create the core
		// FWAnalyzer object.
		return new FWAnalyzer(type, isDistance, frame, word, cache, cacheSize); 
	}
	
	/**
	 * The constructor merely initializes the immutable instance variables to
	 * their values using appropriate parameters.
	 * 
	 * @param type
	 *            The type of f/w analyzer that this object contains information
	 *            about.
	 * @param isDistance If this f/w analyzer provides a distance metric value 
	 * then this flag must be true. Otherwise it is assumed that this f/w analyzer
	 * provides a similarity metric.
	 * @param windowSize
	 *            The frame/window size that was used to analyze EST sequences.
	 * @param wordSize
	 *            The word size that was used to analyze EST sequences. The word
	 *            size represents subdivisions with a frame/window.
	 * @param cacheType
	 *            The type of cache that was used to cache distance/similarity
	 *            metrics generated during MST construction.
	 * @param cacheSize
	 *            The size of each cache entry (per EST) in number of
	 *            distance/similarity metrics that are to be cached to avoid
	 *            having to recompute metrics.
	 */
	public FWAnalyzer(FWAnalyzerType type, boolean isDistance, 
			int windowSize, int wordSize,
			String cacheType, int cacheSize) {
		this.type       = type;
		this.isDistance = isDistance;
		this.windowSize = windowSize;
		this.wordSize   = wordSize;
		this.cacheType  = cacheType;
		this.cacheSize  = cacheSize;
	}
	
	/**
	 * Determine the type of f/w analyzer that this object contains information
	 * about.
	 * 
	 * @return The type of f/w analyzer that was set when this object was 
	 * created (or loaded from an XML configuration file).
	 */
	FWAnalyzerType getType() { return type; }
	
	/**
	 * Determine if this f/w analyzer uses a distance metric or a similarity
	 * metric.
	 * 
	 * @return This method returns true if the f/w analyzer uses a distance
	 * metric.
	 */
	boolean isDistance() { return isDistance; }
	
	/**
	 * The frame/window size that was used to analyze EST sequences.
	 * 
	 * @return This method returns the frame or window size that was
	 * set when this object was created (or loaded from an XML configuration
	 * file).
	 */
	int getFrameSize() { return windowSize; }

	/**
	 * The word size that was used to analyze EST sequences. The word size
	 * represents subdivisions with a frame/window.
	 * 
	 * @return This method returns the word size that was set when this object
	 *         was created (or loaded from an XML configuration file).
	 */
	int getWordSize() { return wordSize; }

	/**
	 * The type of cache that was used to cache distance/similarity metrics
	 * generated during MST construction.
	 * 
	 * @return This method returns the type of cache that was set when this object
	 *         was created (or loaded from an XML configuration file).
	 */
	String getCacheType() { return cacheType; }

	/**
	 * The size of each cache entry (per EST) in number of distance/similarity
	 * metrics that are to be cached to avoid having to recompute metrics.
	 * 
	 * @return This method returns the cache size that was set when this object
	 *         was created (or loaded from an XML configuration file).
	 */
	int getCacheSize() { return cacheSize; }

	/**
	 * Method to marshall the data stored in this object to become part of
	 * a DOM tree element passed in. This method assumes that the element
	 * passed in corresponds to the parent MSTData node in the DOM tree.
	 * 
	 * @param mstData The DOM element corresponding to the "MSTData"
	 * node that contains this entry.
	 */
	public final void marshall(Element mstData) {
		// Create a top-level server entry for this server
		Element analyzer = DOMHelper.addElement(mstData, "FWAnalyzer", null);
		analyzer.setAttribute("metric", (this.isDistance ? "distance" : "similarity"));
		// Add new sub-elements for each value.
		DOMHelper.addElement(analyzer, "Type", type.toString().toLowerCase());
		DOMHelper.addElement(analyzer, "WindowSize", "" + windowSize);
		DOMHelper.addElement(analyzer, "WordSize", "" + wordSize);
		DOMHelper.addElement(analyzer, "CacheType", cacheType);
		DOMHelper.addElement(analyzer, "CacheSize", "" + cacheSize);
	}
	
	/**
	 * Method to marshall the data stored in this object directly to a
	 * XML fragment. The XML fragment is guaranteed to be compatible
	 * with the PEACE work space configuration data. 
	 * 
	 * @param out The stream to which the XML must be serialized.
	 */
	public final void marshall(PrintWriter out) {
		final String Indent = "\t\t\t";
		// Printf format strings for string & numerical elements
		final String STR_ELEMENT = Indent + "\t" + "<%1$s>%2$s</%1$s>\n";
		final String NUM_ELEMENT = Indent + "\t" + "<%1$s>%2$d</%1$s>\n";
		
		// Create a top-level entry
		out.printf("%s<FWAnalyzer metric=\"%s\">\n", Indent,
				(isDistance() ? "distance" : "similarity")); 
		// Add new sub-elements for each value.
		out.printf(STR_ELEMENT, "Type", type.toString().toLowerCase());
		out.printf(NUM_ELEMENT, "WindowSize", windowSize);
		out.printf(NUM_ELEMENT, "WordSize", wordSize);
		out.printf(STR_ELEMENT, "CacheType", cacheType);
		out.printf(NUM_ELEMENT, "CacheSize", cacheSize);
		// Close the element
		out.printf("%s</FWAnalyzer>\n", Indent);
	}
	
	/**
	 * Return the information in the form of a partial PEACE command 
	 * line.
	 * 
	 * This method can be used to obtain the information associated
	 * with this analyzer as a partial PEACE command line.
	 * 
	 * @return Return the information as a command line parameter.
	 */
	public String toCmdLine() {
		final String[] AnalyzerName = {"twopassD2", "d2", "d2zim", "clu"};
		String cmdLine = "--analyzer " + AnalyzerName[type.ordinal()];
		cmdLine += " --frame " + windowSize;
		cmdLine += " --word "  + wordSize;
		cmdLine += " --cacheType " + cacheType;
		cmdLine += " --cache " + cacheSize;
		return cmdLine;
	}
	
	/**
	 * The size of the window or the frame size that was used for analyzing
	 * EST sequences.
	 */
	private final int windowSize;

	/**
	 * The word size that was used for analyzing EST sequences.
	 */
	private final int wordSize;
	
	/**
	 * The type of cache that was used to hold the data distance/similarity
	 * metric when the analysis was conducted.
	 * 
	 */
	private final String cacheType;
	
	/**
	 * The size of each cache entry (per EST) in number of distance/similarity
	 * metrics that are to be cached to avoid having to recompute metrics.
	 */
	private final int cacheSize;
	
	/**
	 * The type of f/w analyzer that this object contains information about.
	 */
	private final FWAnalyzerType type;
	
	/**
	 * Flag to indicate if this f/w analyzer provides distance metrics
	 * (rather than similarity) for identifying similar ESTs.
	 */
	private final boolean isDistance;
}

File: peace/src/clustering/FWAnalyzer.cpp
#ifndef FW_ANALYZER_CPP
#define FW_ANALYZER_CPP

//--------------------------------------------------------------------
//
// This file is part of PEACE.
// 
// PEACE is free software: you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// PEACE is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with PEACE.  If not, see <http://www.gnu.org/licenses/>.
// 
// Miami University makes no representations or warranties about the
// suitability of the software, either express or implied, including
// but not limited to the implied warranties of merchantability,
// fitness for a particular purpose, or non-infringement.  Miami
// University shall not be liable for any damages suffered by licensee
// as a result of using, result of using, modifying or distributing
// this software or its derivatives.
//
// By using or copying this Software, Licensee agrees to abide by the
// intellectual property laws, and all other applicable laws of the
// U.S., and the terms of GNU General Public License (version 3).
//
// Authors:   Dhananjai M. Rao          raodm@muohio.edu
//
//---------------------------------------------------------------------

#include "ArgParser.h"
#include "FWAnalyzer.h"
#include "ResultLog.h"
#include "ESTList.h"

#include <algorithm>
#include <time.h>

FWAnalyzer::FWAnalyzer(const std::string& analyzerName)
    : ESTAnalyzer(analyzerName) {
    // Initialize instance variables
    frameSize = 100;
    wordSize  = 6;
}

FWAnalyzer::~FWAnalyzer() {
    // Nothing to be done
}

void
FWAnalyzer::addCommandLineArguments(ArgParser& argParser) {
    // Let base class add common parameters.
    ESTAnalyzer::addCommandLineArguments(argParser);    
    const ArgParser::ArgRecord CommonArgsList[] = {
        {"--frame", "Frame size (in base pairs)",
         &frameSize, ArgParser::INTEGER},
        {"--word", "Word size (in base pairs)",
         &wordSize, ArgParser::INTEGER},    
        {"", "", NULL, ArgParser::INVALID}
    };
    argParser.addValidArguments(CommonArgsList);
}

bool
FWAnalyzer::initialize() {
    // Validate command-line parameters first.
    if (frameSize < 0) {
        // Necessary parameters not specified.
        std::cerr << getName()
                  << ": Frame size must be greater than zero "
                  << "(use --frame option)\n";
        return false;
    }
    if ((wordSize < 0) || (wordSize > frameSize)) {
        // Something is wrong with word size information.
        std::cerr << getName()
                  << ": Word size (greater than frame size) not specified? "
                  << "(use --word option)\n";
        return false;
    }
    // OK. basic stuff seems fine. Let base class initialize first
    if (!ESTAnalyzer::initialize()) {
        // Base class initialization failed.
        return false;
    }
    // Everything went well.
    return true;    
}

std::string
FWAnalyzer::getFrame(const EST* est, bool start) {
    std::string result(est->getSequence());
    if (start) {
        return result.substr(0, frameSize);
    }
    // Get base pair from end of est sequence.
    return result.substr(result.size() - frameSize);
}

void
FWAnalyzer::dumpEST(ResultLog& log, const EST* est, const bool isReference) {
    const std::string frame = getFrame(est, !isReference);

    if (isReference) {
        const char* start=(htmlLog ? "<font face=\"courier\" color=red>" : "");
        const char* end  =(htmlLog ? "</font>" : "");
        log.report(est->getInfo(), "%s%s%s", "n/a", start, frame.c_str(), end);
    } else {
        const char* start = (htmlLog ? "<font face=\"courier\">" : "");
        const char* end   = (htmlLog ? "</font>" : "");
        log.report(est->getInfo(), "%s%s%s", "%f",
                   start, frame.c_str(), end, est->getSimilarity());
    }
}

void
FWAnalyzer::dumpESTList(const ESTList& estList,
                        const EST* refEST,
                        ResultLog& log) {
    // Start up a Table in the log.
    const char* Titles[] = {"Name", "EST Frame", "Metric", NULL};
    log.startTable(Titles);
    // Dump each est out.
    for(int id = 0; (id < (int) estList.size()); id++) {
        if (id % 20 == 0) {
            // Dump reference EST out.
            dumpEST(log, refEST, true);
        }
        if (estList[id] == refEST) {
            // Don't dump reference EST again;
            continue;
        }
        // Dump a non-Reference EST out.
        dumpEST(log, estList[id], false);
    }
}


int
FWAnalyzer::setReferenceEST(const EST* est) {
    ASSERT( est != NULL );
    // Clear out the previous reference EST information.
    referenceFrame = "";
    // Setup the new reference EST.
    refEST  = est;
    // Obtain list of ESTs for further processing.
    if (refEST->getID() >= estList->size()) {
        // Invalid reference EST id.
        std::cerr << "Reference EST index is greater than number of ESTs.\n"
                  << "Cannot continue further processing.\n";
        return 2;
    }
    
    // Obtain the reference frame from end of the reference EST sequence.
    referenceFrame = getFrame(refEST, false);
    // Return 0 to indicate things proceeded successfully.
    return 0;
}

float
FWAnalyzer::getMetric(const EST* otherEST) {
    ASSERT( otherEST != NULL );
    if (otherEST->getID() == refEST->getID()) {
        // Comparing reference to itself simply results in 0 all the
        // time, in all FWAnalyzers
        return 0;
    }
    return  analyzeFrame(referenceFrame, getFrame(otherEST), wordSize);
}

int
FWAnalyzer::analyze() {
    int result = initialize();
    if (result != 0) {
        // Error occured during initialization...
        return result;
    }
    // Set the reference EST.
    EST* refEST = estList->get(0, true);
    if ((result = setReferenceEST(refEST)) != 0) {
        // Error occured when setting up the reference EST...
        return result;
    }

    // Perform the core analysis and track total similarity to compute
    // mean value at the end of the loop.
    double total = 0;
    for(int id = 0; (id < estList->size()); id++) {
        // Analyze and update similarity value.
        EST* est = estList->get(id, true);
        const float similarity = ESTAnalyzer::analyze(est);
        est->setSimilarity(similarity);
        total += similarity;
    }
    // Compute mean and then compute variance by reiterating over the
    // list and summing up the square of deviations.
    const double mean = total / estList->size();
    double deviations = 0;
    for(int id = 0; (id < estList->size()); id++) {
        if (id != refEST->getID()) {
            const double diff = estList->get(id)->getSimilarity() - mean;
            deviations += (diff * diff);
        }
    }
    const double variance = deviations / estList->size();
    //Sort the resulting ESTs.
    // std::sort(estList.begin(), estList.end(), EST::LessEST());

    // ----------- Start logs and dump header.----------------
    ResultLog log(outputFileName, htmlLog);
    // Dump header information.
    dumpHeader(log, mean, variance);
    // Dump information about each EST.
    dumpESTList(*estList, refEST, log);
    
    // All the processing went on successfully.
    return 0;
}

void
FWAnalyzer::dumpHeader(ResultLog& log, const double mean,
                       const double variance) {
    const char   *HTMLTags[] = {"<b>", "</b>", "<i>", "</i>", "<u>", "</u>"};
    const char   *TextTags[] = {"*  ", "  *", "", "", "_", "_"};
    const char   **Tags      = (htmlLog ? HTMLTags : TextTags);
    const char   *Title      = (htmlLog ? "E S T &nbsp;&nbsp;&nbsp; A N A L Y S I S &nbsp;&nbsp;&nbsp; R E P O R T" : "E S T     A N A L Y S I S     R E P O R T");
    // Get curren time for reporting
    char  now_str[128];
    getTime(now_str);

    // First log some general information one, line after another with
    // some HTML decorations to make text bold or underlined.  In text
    // mode, many of the HTML decorations are absent.
    log.reportLine("");
    log.reportLine("%s %s %s", Tags[0], Title, Tags[1]);
    log.reportLine("Analysis conducted on %s", now_str);
    log.reportLine("%sEST Analyzer used: %s%s", Tags[0], getName().c_str(),
                   Tags[1]);
    log.reportLine("EST data read from file: %s%s%s", Tags[2], "", // estFileName,
                   Tags[3]);
    log.reportLine("");

    // Now report the parameters used for analysis.  This information
    // is very useful for comparisons later on.
    log.report("Frame Size: %d", "  ", "Word Size: %d", frameSize, wordSize);
    log.report("Reference EST index: %d", "  ",
               "Number of ESTs: %d", refEST->getID(), estList->size());
    log.report("Similarity Metric mean: %lf", "  ",
               "Similarity Metric variance: %lf", mean, variance);
    log.endTable();
    log.reportLine("");
}

float 
FWAnalyzer::analyzeFrame(const std::string&,
                         const std::string&,
                         const int) {
    return 0;
}

#endif
